### DISCUSSION TITLE
TS compilation error when exactOptionalPropertyTypes set true

### DISCUSSION THREAD

**Discussion by victorsmirnov** [aws/aws-cdk#23885] [Q&A]
### General Issue

TypeScript compilation fails when exactOptionalPropertyTypes is set to true

### The Question

Compilation fails when [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) is set to true.

Several other modules also have the same error.

```
> tsc

node_modules/aws-cdk-lib/aws-apigateway/lib/restapi.d.ts:226:31 - error TS2420: Class 'RestApiBase' incorrectly implements interface 'IRestApi'.
  Types of property 'latestDeployment' are incompatible.
    Type 'Deployment | undefined' is not assignable to type 'Deployment'.
      Type 'undefined' is not assignable to type 'Deployment'.

226 export declare abstract class RestApiBase extends Resource implements IRestApi {
                                  ~~~~~~~~~~~
```

### CDK CLI Version

2.13.0

### Framework Version

2.13.0

### Node.js Version

16.12.0

### OS

macOS Big Sur

### Language

Typescript

### Language Version

TypeScript (4.5.5)

### Other information

Most likely, replacing `readonly latestDeployment?: Deployment;` with `readonly latestDeployment: Deployment | undefined;` should fix the issue. I can try this as my first issue in the project. Please let me know if I should make a pull request.

**Comment by longtv2222** [comment]
Can confirm that I'm facing the same issue here, please refer https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#exact-optional-property-types to know why this is an issue.

**Comment by mpellegrini** [comment]
Hello, wondering if there is any updates on this issue. My current work around is to explicity cast the concrete class as the specific interface - vpc as IVpc or table as ITable. But for anyone using a later version of Typescript with a default config will be hitting this issue.

**Comment by kenberkeley** [comment]
Same here 😮‍💨

```
Type 'Cluster' is not assignable to type 'ICluster' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
  Types of property 'defaultCloudMapNamespace' are incompatible.
    Type 'INamespace | undefined' is not assignable to type 'INamespace'.
      Type 'undefined' is not assignable to type 'INamespace'.ts(2375)
```

![image](https://user-images.githubusercontent.com/13287948/212787997-4ac79951-bda0-4b5c-a79f-6607d0f1d9ab.png)

***

Temp fix is using `as`

![image](https://user-images.githubusercontent.com/13287948/212789500-f1695c46-4df9-426b-ba94-33710a523534.png)

**Comment by TheRealAmazonKendra** [comment]
This is a TypeScript 4.4 feature and the CDK is on 3.9. At this time we cannot upgrade our TypeScript version so there's no action item we can take here. See https://github.com/aws/aws-cdk-rfcs/pull/373 for more information on this.

Given the lack of actions to be taken at this time, I'm going to convert this to a discussion so that others can find this context.

**Comment by lo1tuma** [comment]
Is there any update on this issue? Seems like the mentioned RFC has been accepted and merged. So what would be the next steps in order to update the types in CDK so they don’t conflict with enabled `exactOptionalPropertyTypes`?

**Comment by rehos** [comment]
What is holding the CDK team back from just adding ` | undefined` for optional fields to support `exactOptionalPropertyTypes`?

**Comment by GavinZZ** [comment]
Hello, I apologize for the lack of update. I'll start a discussion thread with the internal CDK team and update this thread. IMO, I don't see anything blocking the addition of `| undefined` to optional properties. I would like to confirm with the team to make sure there is no unconsidered scenarios.

**Comment by GavinZZ** [comment]
Thanks for reporting this! The issue occurs because `exactOptionalPropertyTypes: true` enforces stricter handling of optional properties, requiring explicit `undefined` in types. 

While this is something CDK could support, fixing it would require extensive changes across the entire codebase, adding `| undefined` to numerous properties and their usages. Given the significant effort required and the availability of workarounds, we are not prioritizing this at the moment as we have higher-impact `p1` tasks without viable workarounds.  

### **Workaround for Defining a New Construct**  

If you're defining a CDK construct and the property requires something like `prop?: Type`, but the inputted value could be `undefined`, you can use the spread operator to conditionally include the property, like below:  

```ts
new SomeL2Construct(stack, 'id', {
  ...myProp && { prop: myProp }
});
```
This ensures that prop is only included when myProp is defined, preventing TypeScript errors caused by `exactOptionalPropertyTypes`.

### For Other Scenarios

If there’s no viable workaround for your specific case, feel free to submit a PR as a temporary fix. 

We also welcome contributions to help resolve these issues across CDK. If anyone is willing to submit a PR to add this flag and help us enable `exactOptionalPropertyTypes` in CDK repo, feel free to ping me for a review. Let us know if you have any questions!

**Reply by rehos** [reply]
@GavinZZ thanks for your reply. Note that the type error issues are not caused by new custom constructs. We have control over them so we add the proper type definitions - no workaround needed. The type errors are caused by consuming the AWS CDK which is lacking the union types `| undefined` for optional properties and function parameters for all constructs including L1.

The issue with L1 constructs should be fixed in the package `@cdklabs/typewriter`. In think there are two places that need to be patched. I can send a pull request for that.

For this aws-cdk library I have a codemod that adds the `| undefined` when needed. The only (minor) issue after applying the codemod I see now is that the line length sometimes exceeds the 150 character limit (from the repos eslint config) and unfortunately linting with auto fix doesn't work because eslint does not fix max-len issues. I can share the codemod in a gist so you can try it out yourself.

**Reply by GavinZZ** [reply]
@rehos Feel free to draft a PR on the typewriter package change and tag me for a review. Appreciate it!

And please also share with me the example you were talking about and I'll take a deeper look.

**Reply by rehos** [reply]
@GavinZZ To give you some insights. Below are some basic examples that you encounter when using the CDK with  `exactOptionalPropertyTypes: true` (Typescript errors are in the comments below the line that gives an error). The workaround is  to explicitly cast them to the expected interface type. But the whole point of using Typescript is type safety, so we don't want to use explicit typecasts all over the place.

```ts
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as pipelines from 'aws-cdk-lib/pipelines';

export class MyStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const source = pipelines.CodePipelineSource.connection('owner/repo', 'main', {
      connectionArn:
        'arn:aws:codestar-connections:us-east-1:123456789012:connection/1234abcd-12ab-34cd-56ef-123456789012',
    });

    new pipelines.CodeBuildStep('Synth', {
      input: source,
      // Type 'CodePipelineSource' is not assignable to type 'IFileSetProducer' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
      // Types of property 'primaryOutput' are incompatible.
      //   Type 'FileSet | undefined' is not assignable to type 'FileSet'.
      //     Type 'undefined' is not assignable to type 'FileSet'.ts(2375)
      commands: ['npm ci', 'npm run build', 'npx cdk synth'],
    });

    const lambdaFunction = new lambda.Function(this, 'Function', {
      runtime: lambda.Runtime.NODEJS_22_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('src'),
    });

    const alias = new lambda.Alias(this, 'Alias', {
      aliasName: 'alias',
      version: lambdaFunction.currentVersion,
      // Type 'Version' is not assignable to type 'IVersion' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
      //   Types of property 'role' are incompatible.
      //     Type 'IRole | undefined' is not assignable to type 'IRole'.
      //       Type 'undefined' is not assignable to type 'IRole'.ts(2375)
    });

    const functionUrl = new lambda.FunctionUrl(this, 'FunctionUrl', {
      authType: lambda.FunctionUrlAuthType.NONE,
      invokeMode: lambda.InvokeMode.RESPONSE_STREAM,
      function: alias,
      // Type 'Alias' is not assignable to type 'IFunction' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
      //   Types of property 'role' are incompatible.
      //     Type 'IRole | undefined' is not assignable to type 'IRole'.
      //       Type 'undefined' is not assignable to type 'IRole'.ts(2375)
    });
  }
}
```

**Reply by amaral-fly** [reply]
Any updates on this issue?
